modules/mono/glue/GodotSharp/GodotSharp/Core/Aabb.cs

```diff
@@ -730,18 +730,27 @@ public Aabb(real_t x, real_t y, real_t z, real_t width, real_t height, real_t de
         /// Converts this <see cref="Aabb"/> to a string.
         /// </summary>
         /// <returns>A string representation of this AABB.</returns>
-        public override readonly string ToString()
-        {
-            return $"{_position}, {_size}";
-        }
+        public override readonly string ToString() => ToString(null, null);
 
         /// <summary>
         /// Converts this <see cref="Aabb"/> to a string with the given <paramref name="format"/>.
         /// </summary>
         /// <returns>A string representation of this AABB.</returns>
-        public readonly string ToString(string format)
+        public readonly string ToString(string format) => ToString(format, null);
+
+        /// <summary>
+        /// Converts this <see cref="Aabb"/> to a string with the given <paramref name="formatProvider"/>.
+        /// </summary>
+        /// <returns>A string representation of this AABB.</returns>
+        public readonly string ToString(IFormatProvider formatProvider) => ToString(null, formatProvider);
+
+        /// <summary>
+        /// Converts this <see cref="Aabb"/> to a string with the given <paramref name="format"/> and <paramref name="formatProvider"/>.
+        /// </summary>
+        /// <returns>A string representation of this AABB.</returns>
+        public readonly string ToString(string format, IFormatProvider formatProvider)
         {
-            return $"{_position.ToString(format)}, {_size.ToString(format)}";
+            return $"{_position.ToString(format, formatProvider)}, {_size.ToString(format, formatProvider)}";
```

Comment:
Woah, I had no clue that even existed!

Reply:
Actually, looking more into it, I disagree with this implementation. The documentation of [NumberGroupSeparator](https://github.com/dotnet/runtime/blob/a30d6c9dc413d9c60bed2d4daecba4552abd7fe9/src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs#L16-L17) specifies it's for digits to the left of the decimal point. That doesn't feel like it's applicable in this context, as the comma acts as a separator in a list format. I have it changed on the repo for now, but I'm considering a revert

Reply:
It's what `System.Numerics` vectors use for separation, so we'll want to be consistent if we are going to be supporting cultures; otherwise, we shouldn't implement `IFormattable` in these types.

There was a similar discussion in the .NET repository about the Vector64/128/256 types. The conclusion was that these types didn't need to implement `IFormattable`.

- https://github.com/dotnet/runtime/issues/11604
- https://github.com/dotnet/coreclr/pull/25343

The current [`Vector256<T>.ToString`](https://github.com/dotnet/runtime/blob/a30d6c9dc413d9c60bed2d4daecba4552abd7fe9/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256_1.cs#L440-L459) implementation actually uses `NumberGroupSeparator` internally, but since it doesn't implement `IFormattable`, it always uses the Invariant Culture (which uses comma: `,`).

Reply:
Though we *will* be supporting cultures by nature of the actual values being formatted by them. The discussions even seemed to agree that `NumberGroupSeparator` wasn't the best choice for the seperation of a list of elements, before ultimately falling back to not bothering with formatting in the first place. The takeaway I got was that they would've opted for hardcoded commas, much like the hardcoded wrappers (`<`, `>`), if they hadn't gone with the overall invariant style. That, and the Vector types in `System.Numerics` look like the only ones that implemented IFormattable, so I don't believe they should be taken as something to copy in this instance

Reply:
What I mean is that if we use a hardcoded separator (a comma), that's not supporting cultures. Not every culture uses commas to separate numbers. Vector64/128/256 don't implement `IFormattable` because those types don't support cultures, their `ToString` implementation always uses the Invariant Culture.

I think you may be suggesting implementing `IFormattable` using comma as a separator and only using the `IFormatProvider` for the numbers, but that's mixing cultures and in my opinion it's more confusing. For example the german culture uses comma for the decimal separator so a Vector2 would look like `(4,2, 4,2)` with commas everywhere, that doesn't look right to me. I also don't know if parenthesis is the right thing to use for every culture.

So, the more I think about it the more I think these types shouldn't implement `IFormattable`. We may want to wait until Vector64/128/256 implement it to see what they do there. It seems there are plans to implement the interface at some point (see https://github.com/dotnet/runtime/pull/90764#discussion_r1344320650).

Reply:
That is largely what I'm saying, yes. And thats because, while I understand that other cultures might use different separators for numbers, I'm not convinced that `NumberGroupSeparator` is the way to go about it. They specify a different separator for currency for instance, and this separator is specified for splitting large numbers to the left of the decimal place. `NumberGroupSeparator`, as far as I can tell, doesn't specify if it should be used as a separator for different *batches* of numbers; just that it separates the portions of a single, large number

I'm fine waiting out and seeing how these other types are implemented, but I wouldn't want to discard `IFormattable` as a whole, especially if `ISpanFormattable` is planned down the road. Something that could be considered as well is [Unity's](https://github.com/Unity-Technologies/UnityCsReference/blob/258abdf1d3dd9c35c4494d7b7acd3f1ce3d9ae84/Runtime/Export/Math/Vector3.cs#L494-L501) implementation of Vector3, which I would imagine many newer Godot C# users would expect. Failing that, we could always just... Add formatting options? Haven't messed with that much before, but I know there's some format string shenanigans in Godot already, so it wouldn't be totally unprecedented

Reply:
I agree with you that `NumberGroupSeparator` may not be the right separator, but it would at least be consistent with the BCL types, so that's why I suggested it.

If we end up deciding that `IFormattable` shouldn't be implemented for these types, then neither should `ISpanFormattable` because that's just the Span version of the same interface (the version 2.0). But I'm not really saying we should discard it altogether, I just think we should wait and see what the BCL does and follow their lead.

I don't think Unity is a good example because their APIs don't usually follow normal .NET conventions. Also, I don't think the repository you linked is open source so I'll refrain from looking at that code.

I don't know what you mean by formatting options and format string shenanigans. But it sounds like it would grow the scope of just implementing `IFormatting` and that would need a proposal and some justification.

Reply:
Hmm, it's true that would expand the scope of `IFormattable`, and consequently the scope of this PR. I'm beginning to see why you've become hesitant. A proposal is probably be warranted for knowing what direction would be preferable

In any case, it'll likely be well worth implementing some sort of testing system into #82955 to output different formatting styles across various `IFormattable` operations. Actually seeing firsthand what kinds of outputs *could* happen with different formatting options would probably go a long way towards making sure the implementations make sense & aren't off-base in some way